### DISCUSSION TITLE
Creating Customer On paddle_billing.subscription.created

### DISCUSSION THREAD

**Discussion by cfprotools** [pay-rails/pay#949] [Q&A] [answered]
I'll be selling access to my SaaS through a Paddle checkout embedded in a ClickFunnels page.  Once a subscription is created during the sign up process Paddle will fire the webhook to the app so that a User, Team, Customer, etc... can be created.  It seems like the Pay gem is expecting that the User and Pay::Customer will be created first and then the Pay::Customer will be passed to Paddle through the passthrough params.  That's not how I'd like this to work.

How best to customize the webhook flow so that I can search for an existing Team (owned by a User) by the email in the webhook and either attach the subscription if the team and user already exist or create a new user and team and then attach them to the PayCustomer?

**Comment by deanpcmad** [comment]
For Paddle Billing it's best to create the customer first, as then any subscriptions/transactions are created for that `Pay::Customer`, which is then linked to your main model, may it be a User, Team, etc. 

What's the reasoning behind why you don't like the current way? Do you not have users login first?

**Reply by cfprotools** [reply]
yes, correct.  The sign up is done outside of the app.  So the user never logs in.  They don't even get a link to the app until after they've subscribed.  Then they'll get an email with the link to login.  This allows me to keep the entire funnel process outside of the app.  That way it's much easier to make changes from a marketing perspective without having to push code changes to the app whenever we want to change the sign page.

**Reply by cfprotools** [reply]
The Paddle checkout is embedded on a sales page in our ClickFunnels page builder account.  Once the checkout is completed, I want the account creation process to be initiated by the webhook received from Paddle for the subscription creation.  We already have one app working this way and I'm rolling out two others that I want to work the same but using the Pay gem instead of handrolled.

**Reply by deanpcmad** [reply]
So how would the Pay Customer be linked to your model? Do you create the User/Team during that setup or is it *just* the Paddle checkout?

I guess we could add a webhook for `customer.created` (and then likely `customer.updated` too) - thoughts @excid3? 
But this Pay Customer would need to be linked some how to your model.

**Reply by cfprotools** [reply]
Ideally, the webhook listener would check for a Pay Customer to be passed in through the passthrough.  If one is there, then attach the subscription to that customer.  If one is not passed in passthrough, then create a new team, user, and Pay Customer and attach the subscription there.

**Reply by deanpcmad** [reply]
Well Pay won't be able to create the team/user because it requires that, before a Pay Customer is created because they are linked

**Reply by cfprotools** [reply]
Gotcha, so if I add a custom webhook processor like here

https://github.com/pay-rails/pay/blob/main/docs/7_webhooks.md#custom-webhook-listeners

will the custom listener fire before the Pay listener?  Can I override completely and build my own and bypass the Pay listener entirely?

**Reply by deanpcmad** [reply]
I'm not entirely sure. I don't think it will completely override the webhook but maybe @excid3 could help with that?

**Reply by cfprotools** [reply]
Yeah, If I could capture the webhook and process with logic to find or create the team, user prior to the Subscription.sync being called, then I think we'd be good.  I just am trying to sort out where to inject into the existing process without having to fork the whole thing.  If the custom webhook handler fires first, then maybe I could do it there.  Of, if it overrides entirely and stops the regular processing, then I could do it there and then call the Subscription.sync call from the custom one.

**Reply by cfprotools** [reply]
ah, I think I might see how to do this now in the docs.  Looks like I can unsubscribe the default listener and then create a new custom listener and subscribe that one instead.  The new listener would check for an existing team, user by email address and create them if needed, then continue on to the normal sync call just like the default listener.

Does that sound right?

**Reply by deanpcmad** [reply, marked as answer]
I've never needed to create custom webhook listeners, but yes it does appear so looking at the docs > https://github.com/pay-rails/pay/blob/main/docs/7_webhooks.md#unsubscribing-from-a-webhook-listener

**Reply by cfprotools** [reply]
Just an updated.  I've got some updates made to override the incoming webhook for new Paddle subscriptions.  I'm still working on testing, but I think I've got the solution I need.

Thanks for all your help.

**Reply by deanpcmad** [reply]
Sounds good! Glad to be of help 😄

**Reply by ninsky** [reply]
> For Paddle Billing it's best to create the customer first, as then any subscriptions/transactions are created for that `Pay::Customer`, which is then linked to your main model, may it be a User, Team, etc.
> 
> What's the reasoning behind why you don't like the current way? Do you not have users login first?

Do you have a source showing how to create a customer first? Having a hard time to integrate Paddle

**Reply by deanpcmad** [reply]
> Do you have a source showing how to create a customer first? Having a hard time to integrate Paddle

Yes, it's all in the docs - https://github.com/pay-rails/pay/blob/main/docs/paddle_billing/1_overview.md
But you would set the payment processor and then use the `api_record` method to create the customer on Paddle
